src/transformers/integrations/ggml.py

```diff
@@ -0,0 +1,428 @@
+        "vocab_size": "vocab_size",
+    },
+    "tokenizer": {
+        "ggml.model": "model_type",
+        "ggml.bos_token_id": "bos_token_id",
+        "ggml.eos_token_id": "eos_token_id",
+        "ggml.unknown_token_id": "unk_token_id",
+        "ggml.padding_token_id": "pad_token_id",
+    },
+}
+
+GGUF_TOKENIZER_MAPPING = {
+    "tokenizer": {
+        "ggml.model": "tokenizer_type",
+        "ggml.tokens": "tokens",
+        "ggml.scores": "scores",
+        "ggml.token_type": "token_type",
+        "ggml.merges": "merges",
+        "ggml.bos_token_id": "bos_token_id",
+        "ggml.eos_token_id": "eos_token_id",
+        "ggml.unknown_token_id": "unk_token_id",
+        "ggml.padding_token_id": "pad_token_id",
+        "ggml.add_space_prefix": "add_prefix_space",
+    },
+    "tokenizer_config": {
+        "chat_template": "chat_template",
+        "ggml.model": "model_type",
+        "ggml.bos_token_id": "bos_token_id",
+        "ggml.eos_token_id": "eos_token_id",
+        "ggml.unknown_token_id": "unk_token_id",
+        "ggml.padding_token_id": "pad_token_id",
+    }
+}
+
+
+def _gguf_parse_value(_value, data_type):
+    if not isinstance(data_type, list):
+        data_type = [data_type]
+    if len(data_type) == 1:
+        data_type = data_type[0]
+        array_data_type = None
+    else:
+        if data_type[0] != 9:
+            raise ValueError("Received multiple types, but therefore expect the first type to indicate an array.")
+        data_type, array_data_type = data_type
+
+    if data_type in [0, 1, 2, 3, 4, 5, 10, 11]:
+        _value = int(_value[0])
+    elif data_type in [6, 12]:
+        _value = float(_value[0])
+    elif data_type in [7]:
+        _value = bool(_value[0])
+    elif data_type in [8]:
+        _value = "".join([chr(a) for a in _value])
+    elif data_type in [9]:
+        _value = _gguf_parse_value(_value, array_data_type)
+    return _value
+
+
+def _gguf_read_value(f, data_type):
+    if data_type == DATA_TYPES["string"]:
+        length = struct.unpack("<Q", f.read(8))[0]
+        return f.read(length).decode("utf-8")
+
+    elif data_type == DATA_TYPES["uint32"]:
+        return struct.unpack("<I", f.read(4))[0]
+
+    elif data_type == DATA_TYPES["uint64"]:
+        return struct.unpack("<Q", f.read(8))[0]
+
+    elif data_type == DATA_TYPES["int32"]:
+        return struct.unpack("<i", f.read(4))[0]
+
+    elif data_type == DATA_TYPES["float32"]:
+        return struct.unpack("<f", f.read(4))[0]
+
+    elif data_type == DATA_TYPES["array"]:
+        data_type, count = struct.unpack("<IQ", f.read(4 + 8))
+        return [_gguf_read_value(f, data_type) for _ in range(count)]
+    elif data_type == DATA_TYPES["bool"]:
+        # This should correspond to `GGUF_METADATA_VALUE_TYPE_BOOL`
+        # 1-byte value where 0 is false and 1 is true.
+        return struct.unpack("<b", f.read(1))[0]
+    else:
+        raise NotImplementedError(f"Data type {data_type} not implemented")
```

Comment:
(younes) IMO this and the method below are doing basically the same thing as `_gguf_parse_value` and `load_gguf_checkpoint_in_pytorch_model`.

We should clean that up